### PR TITLE
Add ComplexInputs to upload local files to Docker containers

### LINKED ISSUE #52: osprey cannot access local input configuration files
**Describe the bug**
`osprey` cannot access local input configuration files since they are passed into the processes as a literal input.

**To Reproduce**
1. Jupyter lab demo for any process
2. Change the relative configuration file path to an absolute file path using `pkg_resources.resource_filename`
3. Run the process

**Expected behavior**
Replace the literal inputs with complex inputs.

### LINKED ISSUE #56: Add ComplexInputs for each of the required files in the configurations
**Is your feature request related to a problem? Please describe.**
Local files are not accessible from Docker containers.

**Describe the solution you'd like**
Add ComplexInputs for each of the required files in the configurations

**Describe alternatives you've considered**
Allow users to upload local files to THREDDS server and use the OPeNDAP URLs for the configruations.

### PR DESCRIPTION
**Pull request by sum1lim**
Resolves #52 
Resolves #56 

`Osprey` processes can now take local inputs when running in docker containers. Inputs required for RVIC processes are added as ComplexInputs, and `Osprey` processes will make a configuration dictionary using the inputs. When testing `jupyter lab demos` or `make`, change the `docker_url` to `http://docker-dev03.pcic.uvic.ca:30124/wps`, which hosts the docker container for this branch.

### CODE REVIEW

**Review by nikola-rados — CHANGES_REQUESTED** [review]
There are a ton of changes here, most of it looks really good. We may want to consider cleaning some things up and improving the readability, but overall this is on the right track.

File: osprey/processes/wps_convert.py [hunk 1]
```diff
@@ -24,13 +24,38 @@ def __init__(self):
...
+                max_occurs=1,
+                supported_formats=[FORMATS.NETCDF, FORMATS.DODS],
+            ),
+            ComplexInput(
+                "config_file",
+                "Convert Configuration",
+                abstract="Path to input configuration file for Convert process (optional)",
+                min_occurs=1,
+                max_occurs=1,
+                supported_formats=[Format("text/cfg", extension=".cfg")],
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/processes/wps_convert.py:57]: In `thunderbird`, yes we were. And it kind of happened in `osprey` as well. So, the strange behaviour was that `vars(request.inputs["updates_file"][0])._file` was shown as `None` when running in docker containers, and I assumed the copy of the input file, `input.extension`, is not being created in `tmp`. This bugged me for a long time. However, with curiosity, when I just tried logging `request.inputs["updates_file"][0].file` to see if it is `None` for sure, it gave a filepath with actual content in it.

In short, we still see some bizarre behaviour when logging out `__dict__` attribute of inputs with a non-predefined format, but we still can _"use"_ the `file` attribute. I will have to make sure if it applies to `thunderbird` as well, and it would be great if it works.

File: osprey/processes/wps_convert.py [hunk 2]
```diff
@@ -68,8 +117,7 @@ def _handler(self, request, response):
             process_step="process",
         )
 
-        tmp_config_file = replace_urls(config_file, self.workdir)
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/processes/wps_convert.py:71]: The functionality of `replace_urls` is merged in `collect_args` so we no longer need it.

File: osprey/processes/wps_convolution.py [hunk 3]
```diff
@@ -27,7 +27,7 @@ def __init__(self):
                 "CASE_DIR": None,
                 "CASEID": None,
                 "CASESTR": "historical",
-                "CALENDAR": None,
+                "CALENDAR": "standard",
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/processes/wps_convolution.py:30]: I think it is a good idea. I don't see any constraints on doing so. How about I make a separate `templates.py` instead of including them in `utils.py`?

[the remark this answers is not in the document]
Comment (nikola-rados) [reply osprey/processes/wps_convolution.py:30]: I like that idea.

File: osprey/processes/wps_convolution.py [hunk 4]
```diff
@@ -75,13 +75,71 @@ def __init__(self):
...
         }
         inputs = [
+            log_level,
             LiteralInput(
-                "convolve_config",
-                "Configuration",
-                abstract="Path to input configuration file or input dictionary",
+                "case_id",
+                "Case ID",
+                abstract="Case ID for the RVIC process (required)",
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/processes/wps_convolution.py:82]: Aren't these abstracts shown in `bird?`. I thought it would give a better direction of using the birds.

File: osprey/processes/wps_convolution.py [hunk 5]
```diff
@@ -99,8 +157,50 @@ def __init__(self):
             status_supported=True,
         )
 
+    def config_handler(self, workdir, args):
+        if "convolve_config_file" in args:
+            unprocessed = read_config(args["convolve_config_file"])
+        elif "convolve_config_dict" in args:
+            unprocessed = eval(args["convolve_config_dict"])
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/processes/wps_convolution.py:164]: Yeah, it's to convert `dict`-like strings to an actual `dict`

File: osprey/processes/wps_convolution.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (nikola-rados) [reply osprey/processes/wps_convolution.py:82]: Yes they are shown and yes it would probably provide more direction. I was just pointing out that the bird would complain if there was any item missing.

File: osprey/utils.py [hunk 6]
```diff
@@ -117,3 +49,26 @@ def get_outfile(config, dir_name):
     (out_file,) = collect_output_files(filename, outdir)
 
     return os.path.join(outdir, out_file)
+
+
+def collect_args(request, workdir):
```
[the remark this answers is not in the document]
Comment (sum1lim) [reply osprey/utils.py:54]: > I still feel we could assign individual items here, or keep the dict and just deconstruct it upon returning.

Returning the `dict` and deconstructing it after the function call sounds better for me.

File: osprey/utils.py [hunk 7]
```diff
@@ -161,7 +161,19 @@ def convolve_config_handler(
...
+def rvic_config_validator(cfg):
+    return {
+        section: {
+            key: cfg[section][key]
+            if type(cfg[section][key]) != list or len(cfg[section][key]) > 1
+            else cfg[section][key][0]
+            for key in cfg[section].keys()
+        }
+        for section in cfg.keys()
+    }
```
Comment (sum1lim) [inline osprey/utils.py:179]: @nikola-rados Surprisingly, storing `config_template`s in a separate file caused a problem, and it is also very odd. Only dictionary values in the `HISTORY` section in `convolve_config_template` are changed when the dictionary is imported in `wps_convolution` and `wps_full_rvic`. This function `rvic_config_validator` is to re-convert the values to they original data types.

File: osprey/config_templates.py [hunk 8]
```diff
@@ -0,0 +1,105 @@
+from collections import OrderedDict
+
+params_config_template = OrderedDict(
```
Comment (nikola-rados) [inline osprey/config_templates.py:3]: Thanks for moving these in here. Is there a reason why you decided to go with `OrderedDict`?

File: osprey/processes/wps_convert.py [hunk 9]
```diff
@@ -47,8 +73,31 @@ def __init__(self):
...
+        processed = ".".join(unprocessed.split(".")[:-1]) + "_edited.cfg"
+        with open(processed, "w") as cfg:
+            parser.write(cfg)
+
+        return processed
+
     def _handler(self, request, response):
-        loglevel = request.inputs["loglevel"][0].data
+        args = collect_args(request, self.workdir)
+        loglevel, uhs_files, station_file, domain, config_file = tuple(args.values())
```
Comment (nikola-rados) [inline osprey/processes/wps_convert.py:99]: While this does solve the comment I made above, we are straining things a bit harder than we should. Instead of turning the values from `args` into a tuple, could we not just make `args` a tuple in the first place?

**Review by nikola-rados — APPROVED** [review]
Just a couple tiny comments but I think this is good to go. Nice work!

File: osprey/config_templates.py (hunk 8 above)
Comment (sum1lim) [reply osprey/config_templates.py:3]: Functionally, there isn't really any difference in returning `dictionary` or `OrderedDict`. However, the `read_config` function of RVIC returns `OrderedDict` and I wanted to make sure `config_handler` is returning the same type.

File: osprey/processes/wps_convert.py (hunk 9 above)
Comment (sum1lim) [reply osprey/processes/wps_convert.py:99]: It is challenging because the size of `collect_args`'s return value differs depending on the optional inputs. Since I chose to take the advantage of dynamically defined args, I think this is the best idea. Perhaps you have any suggestions?

Comment (nikola-rados) [reply osprey/processes/wps_convert.py:99]: That is a fair point. I think it is fine then, it may be helpful to condense the code into a single like if you can though. Otherwise I think it should be ok.
